python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+        return self.host_buffer.get().size()
+
+    def __dealloc__(self):
+        if not self._freed:
+            CudaDeviceManager().free_host(self)
+
+
+cdef class CudaBufferReader(NativeFile):
+    """File interface for zero-copy read from CUDA buffers.
+
+    Note: Read methods return pointers to device memory. This means
+    you must be careful using this interface with any Arrow code which
+    may expect to be able to do anything other than pointer arithmetic
+    on the returned buffers.
+    """
+    def __cinit__(self, CudaBuffer obj):
+        self.buffer = obj
+        self.reader = new CCudaBufferReader(self.buffer.buffer)
+        self.rd_file.reset(self.reader)
+        self.is_readable = True
+        self.closed = False
+
+    def read_buffer(self, nbytes=None):
+        """Read into new device buffer.
+
+        Parameters
+        ----------
+        nbytes : int, default None
+          Specify the number of bytes to read. Default: None (read all
+          remaining bytes).
+
+        Returns
+        -------
+        cbuf : CudaBuffer
+          New device buffer.
+        """
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_read = 0
+            shared_ptr[CCudaBuffer] output
+
+        if nbytes is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = nbytes
+
+        with nogil:
+            check_status(self.reader.Read(c_nbytes,
+                                          < shared_ptr[CBuffer] * > & output))
+
+        return pyarrow_wrap_cudabuffer(output)
+
+
+cdef class CudaBufferWriter(NativeFile):
+    """File interface for writing to CUDA buffers, with optional buffering.
+    """
+    def __cinit__(self, CudaBuffer buffer):
+        self.buffer = buffer
+        self.writer = new CCudaBufferWriter(self.buffer.cuda_buffer)
+        self.wr_file.reset(self.writer)
+        self.is_writable = True
+        self.closed = False
+
+    def writeat(self, int64_t position, object data):
+        """Write data to buffer starting from position.
+
+        Parameters
+        ----------
+        position : int
+          Specify device buffer position where the data will be
+          written.
+        data : array-like
+          Specify data, the data instance must implement buffer
+          protocol.
+        """
+        if isinstance(data, six.string_types):
+            data = tobytes(data)
+
+        cdef Buffer arrow_buffer = py_buffer(data)
+
+        cdef const uint8_t * buf = arrow_buffer.buffer.get().data()
+        cdef int64_t bufsize = len(arrow_buffer)
+        with nogil:
+            check_status(self.writer.WriteAt(position, buf, bufsize))
+
+    def flush(self):
+        """ Flush the buffer stream """
+        with nogil:
+            check_status(self.writer.Flush())
+
+    def seek(self, int64_t position, int whence=0):
+        # TODO: remove this method after NativeFile.seek supports
+        # writeable files.
```

Comment:
OK, we should probably undertake this refactoring shortly after this patch goes in

Reply:
That's ARROW-3189